### DISCUSSION TITLE
Help typing a not-so-generic combinator

### DISCUSSION THREAD

**Discussion by basro** [winnow-rs/winnow#416] [Q&A] [answered]
I want to write a custom combinator that is not generic over the input and output type but I'm having a lot of trouble writing the output type.

What I wanted to do:
```rust
fn example_combinator(prefix: &'static str) -> impl Parser<_, _, _> // What types go here?
{
  (prefix, take_while(0.., '_')).recognize()
}
```

As a work around I can do this:
```rust
fn example_combinator2<'s>(prefix: &'static str) -> impl Fn(&mut &'s str) -> PResult<&'s str> {
    move |input| (prefix, take_while(0.., '_')).recognize().parse_next(input)
}
```

**Comment by epage** [comment, marked as answer]
```rust
fn example_combinator(prefix: &'static str) -> impl Parser<&str, &str, ContextError>` // What types go here?
{
  (prefix, take_while(0.., '_')).recognize()
}
```
With `Parser<I, O, E>`
- `I` maps to the input to your `Fn`
- `O` maps to the first (or only) parameter for `PResult`
- `E` maps to the second parameter to `PResult` which is defaulted to be `ContextError`

**Comment by basro** [comment]
Thanks for your answer and this great library!
